zjit/src/codegen.rs

```diff
@@ -979,6 +979,16 @@ fn gen_guard_type(jit: &mut JITState, asm: &mut Assembler, val: lir::Opnd, guard
         // Check if opnd is Fixnum
         asm.test(val, Opnd::UImm(RUBY_FIXNUM_FLAG as u64));
         asm.jz(side_exit(jit, state)?);
+    } else if let Some(expected_class) = guard_type.runtime_exact_ruby_class() {
+        // Guard for exact class types (e.g., ClassExact, ArrayExact, StringExact, etc.)
+        asm_comment!(asm, "guard exact class");
+
+        // Get the class of the value
+        let klass = asm.ccall(rb_yarv_class_of as *const u8, vec![val]);
```

Comment:
If we can inline this somehow, that would be great (needs a check if it's a special const, then load `RBASIC(obj)->klass`)